refactor(PXE): sync from PXE instead of from contracts

mverzilli opened the pull request:
Moves the responsibility of deciding when to trigger a contract sync from Aztec.nr to PXE.

- Adds invocations to `sync_private_state` from PXE at a few key points: before running a private function, before running a nested private function, before running a utility function
- Removes generated calls to trigger message discovery from Aztec.nr
- I had to fix this test because it seems by eliminating message discovery code we freed up some fields? Mentioning in case I'm missing something because I wasn't expecting this side-effect: https://github.com/AztecProtocol/aztec-packages/pull/19193/changes#diff-0756560a0ac2c335e95f35e46503040e5aaf4efea1479af4bdb18536c85d360aR58
- Adds guardrails to prevent `sync_private_state` from being invoked unless it's from PXE itself, to ensure consistent and controlled synchronization points
- As far as I can tell, protocol contracts don't need message discovery so added checks to avoid triggering sync for them
- As far as I understand, this also closes #12553 

Closes F-162

mverzilli commented on noir-projects/aztec-nr/aztec/src/macros/internals_functions_generation/external/private.nr:
```diff
@@ -120,16 +119,6 @@ pub(crate) comptime fn generate_private_external(f: FunctionDefinition) -> Quote
         }
     };
 
-    // All private functions perform message discovery, since they may need to access notes. This is slightly
-    // inefficient and could be improved by only doing it once we actually attempt to read any. Note that the message
-    // discovery call syncs private events as well. We do not sync those here if there are no notes because we don't
-    // have an API that would access events from private functions.
```
TODO: check that we're fine not handling this case with this new implementation

mverzilli replied:
Checked: this is indeed an optimization, if the contract doesn't have any notes, discovery isn't necessary. We should discuss if it's worth keeping it now that we control this process from TS world

nventuro replied:
i don´t think we can easily know from ts that a contract does not have any notes. i´d just do it always i think

Thunkar reviewed (dismissed):
lgtm!

nventuro commented on yarn-project/end-to-end/src/e2e_multiple_blobs.test.ts:
```diff
@@ -54,15 +54,19 @@ describe('e2e_multiple_blobs', () => {
       await broadcastFunction(privateFunctions[1]),
+      await broadcastFunction(privateFunctions[2]),
+      await broadcastFunction(privateFunctions[3]),
+      await broadcastFunction(privateFunctions[4]),
```
This looks odd. I have no context on what this test does, but it does not seem obviously correct to see this side effect. We should investigate.

mverzilli replied:
Here's what I understand:

This test tries to generate a block that results in more than one blob =>
We removed message discovery code from functions =>
Functions got leaner in terms of bytecode size =>
Broadcasting 2 functions is no longer enough to exceed 1 blob =>
Adding 3 more broadcasts is what it takes to re-create this "1 block, 2 blobs" scenario

nventuro reviewed (approved):
Fantastic, thanks!

mverzilli committed:
refactor(PXE): sync from PXE instead of from contracts

Moves the responsibility of deciding when to trigger a contract sync from Aztec.nr to PXE.

- Adds invocations to `sync_private_state` from PXE at a few key points: before running a private function, before running a nested private function, before running a utility function
- Removes generated calls to trigger message discovery from Aztec.nr
- I had to fix this test because it seems by eliminating message discovery code we freed up some fields? Mentioning in case I'm missing something because I wasn't expecting this side-effect: https://github.com/AztecProtocol/aztec-packages/pull/19193/changes#diff-0756560a0ac2c335e95f35e46503040e5aaf4efea1479af4bdb18536c85d360aR58
- Adds guardrails to prevent `sync_private_state` from being invoked unless it's from PXE itself, to ensure consistent and controlled synchronization points
- As far as I can tell, protocol contracts don't need message discovery so added checks to avoid triggering sync for them
- As far as I understand, this also closes #12553

Closes F-162

benesjan commented on yarn-project/pxe/src/storage/contract_store/contract_store.ts:
```diff
@@ -316,4 +317,23 @@ export class ContractStore {
+    // Protocol contracts don't have private state to sync
+    if (!isProtocolContract(contractAddress)) {
+      const syncPrivateStateFunctionCall = await this.getFunctionCall('sync_private_state', [], contractAddress);
+      if (functionToInvokeAfterSync && functionToInvokeAfterSync.equals(syncPrivateStateFunctionCall.selector)) {
+        throw new Error(
+          'Forbidden `sync_private_state` invocation. `sync_private_state` can only be invoked by PXE, manual execution can lead to inconsistencies.',
+        );
+      }
```
I don't see how this check can be relevant given that utility functions are not (currently) composable and private functions cannot (currently) call utility functions.

So currently the only way how someone could manually trigger it is if he manually called `pxe.simulateUtility(...)` but there we use job queue so it would just run the sync twice which doesn't seem to be a problem.

Am I missing something?
